### ISSUE TITLE
MSRV-aware `std_instead_of_core`

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#13158]
### Summary

Items sometimes move from `std` to `core`. For example, [`*UnwindSafe` have moved](https://github.com/rust-lang/rust/pull/84662), `std::error::Error` [will move in Rust 1.81.0](https://github.com/rust-lang/rust/pull/125951), and `std::io::Error` might someday.

When this happens, anyone who uses clippy from nightly or beta, or newer than their MSRV, will get lints they cannot satisfy. Therefore, it would be useful if `std_instead_of_core`, `alloc_instead_of_core`, and `std_instead_of_alloc` were MSRV-aware and only fired if the item in question is stable in `core` or `alloc` at the declared MSRV.

### Lint Name

std_instead_of_core

### Reproducer

I tried this code:

```rust
#![feature(custom_inner_attributes)]
#![warn(clippy::std_instead_of_core)]
#![clippy::msrv = "1.80.0"]

use std::error::Error;
```

I saw this happen:

```
warning: used import from `std` instead of `core`
 --> src/lib.rs:5:5
  |
5 | use std::error::Error;
  |     ^^^ help: consider importing the item from `core`: `core`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#std_instead_of_core
note: the lint level is defined here
 --> src/lib.rs:2:9
  |
2 | #![warn(clippy::std_instead_of_core)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I expected to see this happen: No warning

### Version

```text
rustc 1.82.0-nightly (c1a6199e9 2024-07-24)
binary: rustc
commit-hash: c1a6199e9d92bb785c17a6d7ffd8b8b552f79c10
commit-date: 2024-07-24
host: x86_64-apple-darwin
release: 1.82.0-nightly
LLVM version: 18.1.7
```

### Additional Labels

@rustbot label -C-bug +C-enhancement

**Comment by kpreid** [comment]
Previous related work: #9545 added checking if the proposed path was *unstable*.

**Comment by rktaxe** [comment]
> Previous related work: [#9545](https://github.com/rust-lang/rust-clippy/pull/9545) added checking if the proposed path was _unstable_.

It seems that that isn't entirely effective as this lint is now firing on the unstable `core::io` work tracked at https://github.com/rust-lang/rust/issues/154046.  It's perfectly correct for this to fire if `#![feature(core_io)]` *is* enabled, of course.

**Comment by bushrat011899** [comment]
> It seems that that isn't entirely effective as this lint is now firing on the unstable `core::io` work tracked at [rust-lang/rust#154046](https://github.com/rust-lang/rust/issues/154046). It's perfectly correct for this to fire if `#![feature(core_io)]` _is_ enabled, of course.

Having a look to see if I can fix that. I think it might be a consequence of this aspect of the `std_instead_of_core.rs` test:

```rust
// Some intrinsics are usable on stable but live in an unstable module, but should still suggest
// replacing std -> core
fn intrinsic(a: *mut u8, b: *mut u8) {
    unsafe {
        std::intrinsics::copy(a, b, 1);
        //~^ std_instead_of_core
    }
}
```

EDIT: Hunch was correct, have a fix in rust-lang/rust#156164.
